### DISCUSSION TITLE
How to improve CPU utilization

### DISCUSSION THREAD

**Discussion by Leadoc1996** [Anura3D/Anura3D_OpenSource#22] [Q&A] [answered]
Anura3D is not computationally efficient,How to improve

**Comment by murphy-john** [comment]
Hi Leadoc1996, will you expand on your question?

**Reply by Leadoc1996** [reply]
My computer has sixteen cores.But Anura3D uses only one core.I know that Anura3D can do multi-core computing.How do I make it happen. I did not find any information on this issue in the manual.Is it necessary to modify the Anura3D source code?Could you give me some ideas? Thank you  so much.

**Reply by Leadoc1996** [reply]
I really need your help

**Reply by murphy-john** [reply]
Hi Leadoc,
Unfortunately, Anura3d doesn't support multicore computing. There was an attempt to use OpenMPI years ago. The developers at the time could not get consistent results between models run on different numbers of cores and the effort was abandoned. I think there is a strong interest among the Anura 3D community to increase the speed of the code by parallelization, improve efficiency, etc. Most of the developers have background in Geotechnical Engineering and aren't as proficient with computer science as we'd like to be.

**Reply by Leadoc1996** [reply]
All right, but thank you very much

**Reply by murphy-john** [reply]
Yes, we would be very welcoming if you have interest in parallelization and ability to do the implementation.

**Reply by Leadoc1996** [reply]
I hope so. But now I have a new problem.My particle will run out of the boundary, and the calculation will stop. and do you have any idea what might have caused that

**Reply by murphy-john** [reply]
This can be caused by a number of things. The best thing to do is to run the code in debug mode in Visual Studio. 

Double check your boundary conditions. You can also try reducing the courant number.

**Reply by Leadoc1996** [reply]
Or does it have anything to do with my grid

**Reply by murphy-john** [reply]
It's difficult to say without debugging the code. Have you tried that? It's something you'll have to learn to do to understand issues like this. 

From the image below, it does look like there is an "explosion" of MP's. It may be caused by the material parameters or the grid. Again, you must debug the code to find out.

**Reply by Tom-32** [reply]
I want to know if the problem of using multi-core computing is solved in 2024

**Comment by Leadoc1996** [comment]


**Reply by murphy-john** [reply, marked as answer]
Unfortunately, it's not possible to understand what's going on with your model from a single Paraview image. You haven't said what you're trying to model, the features you're using, the model parameters, etc. 

Also, it  appears that your model is very large in terms of number of material points. As is the case with any numerical model, I'd recommend with as simple and small of a model as possible then, only after understanding how you model behaves, adding size and complexity step by step and very slowly.

**Reply by CPLA81192** [reply]
I would like to know if parallel computing is supported

**Comment by elbarrendero** [comment]
@murphy-john  So you compiled a version with OPENMP (Multiprocessing) and got too much different results from single-core processing?

I did some research on the anura3d deepwiki (https://deepwiki.com/search/i-need-more-information-on-how_157ca0e5-0898-4017-9d5d-b9f347dd08c1?mode=deep ), and it seems we could implement the multiprocessing, although i think the issue on the difference in results may be given to the nature of the physics... I really would like to implement MP.
